packages/replay/src/coreHandlers/util/fetchUtils.ts

```diff
@@ -166,3 +183,63 @@ function _getFetchRequestArgBody(fetchArgs: unknown[] = []): RequestInit['body']
 
   return (fetchArgs[1] as RequestInit).body;
 }
+
+function getAllHeaders(headers: Headers, allowedHeaders: string[]): Record<string, string> {
+  const allHeaders: Record<string, string> = {};
+
+  allowedHeaders.forEach(header => {
+    if (headers.get(header)) {
+      allHeaders[header] = headers.get(header) as string;
+    }
+  });
+
+  return allHeaders;
+}
+
+function getRequestHeaders(fetchArgs: unknown[], allowedHeaders: string[]): Record<string, string> {
+  if (fetchArgs.length === 1 && typeof fetchArgs[0] !== 'string') {
+    return getHeadersFromOptions(fetchArgs[0] as Request | RequestInit, allowedHeaders);
+  }
+
+  if (fetchArgs.length === 2) {
+    return getHeadersFromOptions(fetchArgs[1] as Request | RequestInit, allowedHeaders);
+  }
+
+  return {};
+}
+
+function getHeadersFromOptions(
+  input: Request | RequestInit | undefined,
+  allowedHeaders: string[],
+): Record<string, string> {
+  if (!input) {
+    return {};
+  }
+
+  const headers = input.headers;
+
+  if (!headers) {
+    return {};
+  }
+
+  if (headers instanceof Headers) {
+    return getAllHeaders(headers, allowedHeaders);
+  }
+
+  // We do not support this, as it is not really documented (anymore?)
+  if (Array.isArray(headers)) {
+    return {};
+  }
+
+  // Ensure we have lower cased keys & keys that are allowed
+  const filteredHeaders: Record<string, string> = {};
+
+  Object.keys(headers).forEach(key => {
+    const normalizedKey = key.toLowerCase();
+    if (allowedHeaders.includes(normalizedKey)) {
+      filteredHeaders[normalizedKey] = headers[key];
+    }
+  });
```

Comment:
l: Seems to me like we're doing the same thing here as in XHR but a little differently. Wdyt about either using `reduce` in both cases or `forEach`? Or could we even use `getAllowedHeaders` for both? Might make things a little more unified overall. But logaf-l so feel free to leave it as is!

Reply:
good point!